src/app-router.js

```diff
@@ -40,53 +34,37 @@ export class AppRouter extends Router {
     });
   }
 
-  dequeueInstruction() {
-    if (this.isNavigating) {
-      return;
-    }
-
-    var instruction = this.queue.shift();
-    this.queue = [];
-
-    if (!instruction) {
-      return;
-    }
-
-    this.isNavigating = true;
-    this.events.publish('router:navigation:processing', instruction);
-
-    var context = this.createNavigationContext(instruction);
-    var pipeline = this.pipelineProvider.createPipeline(context);
+  dequeueInstruction(instructionCount = 0) {
+    return Promise.resolve().then(() => {
+      if (this.isNavigating && !instructionCount) {
```

Comment:
Sounds good. That was a bit wonky before. I like the consistency you are adding to the design. Keep it up!